cmd/build/v2/smartbuild/hasher.go

```diff
@@ -11,58 +11,67 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package v2
+package smartbuild
 
 import (
 	"crypto/sha256"
 	"encoding/hex"
 	"fmt"
-	"os"
 	"strings"
+	"sync"
 
-	"github.com/okteto/okteto/pkg/env"
 	oktetoLog "github.com/okteto/okteto/pkg/log"
 	"github.com/okteto/okteto/pkg/model"
-)
-
-const (
-	// OktetoSmartBuildUsingContextEnvVar is the env var to enable smart builds using the build context instead of the project build
-	OktetoSmartBuildUsingContextEnvVar = "OKTETO_SMART_BUILDS_USING_BUILD_CONTEXT"
+	"github.com/spf13/afero"
 )
 
 type repositoryCommitRetriever interface {
 	GetSHA() (string, error)
 	GetLatestDirCommit(string) (string, error)
+	GetDiffHash(string) (string, error)
 }
 
 type serviceHasher struct {
 	gitRepoCtrl repositoryCommitRetriever
+	fs          afero.Fs
 
 	buildContextCache map[string]string
+	diffHashCache     map[string]string
 	projectCommit     string
+
+	// lock is a mutex to provide thread safety
+	lock sync.RWMutex
 }
 
-func newServiceHasher(gitRepoCtrl repositoryCommitRetriever) *serviceHasher {
+func newServiceHasher(gitRepoCtrl repositoryCommitRetriever, fs afero.Fs) *serviceHasher {
 	return &serviceHasher{
 		gitRepoCtrl:       gitRepoCtrl,
 		buildContextCache: map[string]string{},
+		diffHashCache:     map[string]string{},
+		fs:                fs,
 	}
 }
 
 // hashProjectCommit returns the hash of the repository's commit
-func (sh *serviceHasher) hashProjectCommit(buildInfo *model.BuildInfo) string {
-	if sh.projectCommit == "" {
+func (sh *serviceHasher) hashProjectCommit(buildInfo *model.BuildInfo) (string, error) {
+	sh.lock.Lock()
+	projectCommit := sh.projectCommit
+	sh.lock.Unlock()
```

Comment:
why we need a lick here? when could be used at the same time?

Reply:
We need it when we are checking if the `GetServicesToBuild`. In order to get the services that are already built. We check that the services are built in parallel. If we don't have the lock there is a race condition and we would be computing the project commit at least two times